From e955f0c983403e8c813cc59d152fa3889c68fa10 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:35:06 -0400
Subject: [alsa] hda: verify pin-converter connection on unsol event for HSW and VLV

Message-id: <140972823573-335-patchset-jkysela@redhat.com>
Patchwork-id: 91031
O-Subject: [RHEL 7.1 PATCH 334/442] ALSA: hda - verify pin:converter connection on unsol event for HSW and VLV
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

This patch will verify the pin's coverter selection for an active stream
when an unsol event reports this pin becomes available again after a display
mode change or hot-plug event.

For Haswell+ and Valleyview: display mode change or hot-plug can change the
transcoder:port connection and make all the involved audio pins share the 1st
converter. So the stream using 1st convertor will flow to multiple pins
but active streams using other converters will fail. This workaround
is to assure the pin selects the right conveter and an assigned converter is
not shared by other unused pins.

Signed-off-by: Mengdong Lin <mengdong.lin@intel.com>
Cc: <stable@vger.kernel.org>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: b4f75aea553a2146bbdd159c397a2ac42cbb9902)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/patch_hdmi.c b/sound/pci/hda/patch_hdmi.c
index be0a9ee..3e4417b 100644
--- a/sound/pci/hda/patch_hdmi.c
+++ b/sound/pci/hda/patch_hdmi.c
@@ -1594,10 +1594,18 @@ static bool hdmi_present_sense(struct hdmi_spec_per_pin *per_pin, int repoll)
    * Re-setup pin and infoframe. This is needed e.g. when
    * - sink is first plugged-in (infoframe is not set up if !monitor_present)
    * - transcoder can change during stream playback on Haswell
+   *   and this can make HW reset converter selection on a pin.
    */
-  if (eld->eld_valid && !old_eld_valid && per_pin->setup)
+  if (eld->eld_valid && !old_eld_valid && per_pin->setup) {
+   if (is_haswell_plus(codec) || is_valleyview(codec)) {
+    intel_verify_pin_cvt_connect(codec, per_pin);
+    intel_not_share_assigned_cvt(codec, pin_nid,
+       per_pin->mux_idx);
+   }
+
    hdmi_setup_audio_infoframe(codec, per_pin,
          per_pin->non_pcm);
+  }
  }
 
  if (eld_changed)
-- 
1.7.1